clmul/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+//!
+//! For optimal performance, set `target-cpu` in `RUSTFLAGS` to `sandybridge`
+//! or newer:
+//!
+//! Example:
+//!
+//! ```text
+//! $ RUSTFLAGS="-Ctarget-cpu=sandybridge" cargo bench
+//! ```
+
+#![cfg_attr(not(test), no_std)]
+#![cfg_attr(all(feature = "armv8", target_arch = "aarch64"), feature(stdsimd))]
+
+mod backend;
+pub use crate::backend::Clmul;
+
+#[cfg(test)]
+mod tests {
+    use cfg_if::cfg_if;
+    use rand::Rng;
+    use rand_chacha::rand_core::SeedableRng;
+    use rand_chacha::ChaCha12Rng;
+
+    cfg_if! {
+        if #[cfg(all(any(target_arch = "x86_64", target_arch = "x86")))] {
+            #[path = "../backend/clmul.rs"]
+            mod clmul;
+            use std::arch::x86_64::*;
+
+            /// Carryless multiplication. Reference implementation.
+            ///
+            /// This code was adapted from swanky
+            /// https://github.com/GaloisInc/swanky/blob/ac7d5d1e8286bbcddcdaf5501d5d925fe79d0591/scuttlebutt/src/block.rs#L51
+            /// which in turn adapted it from the EMP toolkit's implementation.
+            fn clmul128(a: u128, b: u128) -> (u128, u128) {
+                unsafe {
+                    let x = std::mem::transmute(a);
+                    let y = std::mem::transmute(b);
+                    let zero = _mm_clmulepi64_si128(x, y, 0x00);
+                    let one = _mm_clmulepi64_si128(x, y, 0x10);
+                    let two = _mm_clmulepi64_si128(x, y, 0x01);
+                    let three = _mm_clmulepi64_si128(x, y, 0x11);
+                    let tmp = _mm_xor_si128(one, two);
+                    let ll = _mm_slli_si128(tmp, 8);
+                    let rl = _mm_srli_si128(tmp, 8);
+                    let x = _mm_xor_si128(zero, ll);
+                    let y = _mm_xor_si128(three, rl);
+                    (std::mem::transmute(x), std::mem::transmute(y))
+                }
+            }
+        }
+        else if #[cfg(all(target_arch = "aarch64", feature = "armv8"))] {
+            #[path = "../backend/pmull.rs"]
+            mod pmull;
+        }
+    }
+
+    // TODO I had to create an empty tests folder otherwise the path
+    // tests/../backend/soft32.rs was not found. Is this expected???
+    #[path = "../backend/soft32.rs"]
+    mod soft32;
+    #[path = "../backend/soft64.rs"]
+    mod soft64;
+
+    #[test]
+    fn clmul_test() {
+        use soft32::Clmul as s32;
+        use soft64::Clmul as s64;
+
+        let mut rng = ChaCha12Rng::from_entropy();
+        let a: [u8; 16] = rng.gen();
+        let b: [u8; 16] = rng.gen();
+
+        let (r64_0, r64_1) = s64::new(&a).clmul(s64::new(&b));
+        let (r32_0, r32_1) = s32::new(&a).clmul(s32::new(&b));
+        assert_eq!(u128::from(r64_0), u128::from(r32_0));
+        assert_eq!(u128::from(r64_1), u128::from(r32_1));
+
+        cfg_if! {
+            if #[cfg(all(any(target_arch = "x86_64", target_arch = "x86")))] {
+                use clmul::Clmul as clm;
```

Comment:
Why are you branching on the backend in `lib.rs`? If you import `backend::Clmul` it should just do the right thing